core/generator.ts

```diff
@@ -37,7 +37,7 @@ export type BlockGenerator = (
 /**
  * Class for a code generator that translates the blocks into a language.
  */
-export class CodeGenerator {
+export abstract class CodeGenerator {
```

Comment:
This is definitely a breaking change, and in particular one that makes the current state of the custom code generator codelab invalid (see google/blockly-samples#1858) as it creates an instance of `CodeGenerator` directly.  (And same for any developer-written custom code generators not updated with the changes announced at this year's summit.)

Although I think subclassing `CodeGenerator` is best practice, I don't think we should force developers to do so.

Reply:
The `abstract` keyword only affects people using typescript right? They shouldn't be cludging on properties anyway, so I don't think this should affect them. (e.g. assigning directly to `scrub_` wouldn't work in typescript because it's protected)

I think we should force people to subclass `CodeGenerator` in the future, because it actually isn't useable on its own. At least as long as it's not breaking peeps more than we are anyway by (eventually) actually exporting types for our generators.

I may be misunderstanding what `abstract` does though, or the current state of how people are using `CodeGenerator` in the wild.

Reply:
Discussed with Rachel and agreed that if you also fix the codelab and any documentation, it's reasonable to make this `abstract`. Since there's no way for typescript developers to be using this with correct types anyway, we also don't need to do a major version bump. We should call this out in release notes though / make sure the PR description explains the change.

I didn't discuss this with Rachel though so you can push back if you disagree, but I do think it might make sense to hold that particular change until we publish usable types for the generators we provide. That way TS devs can make all of their necessary changes at once.

Reply:
Filed an issue for this (https://github.com/google/blockly/issues/7401). Didn't mark it as `triage` since you already discussed. Linked to https://github.com/google/blockly/issues/6828